File: Shortest Path/Bellman-Ford/1865번 - 웜홀/1865.py
import sys

input = sys.stdin.readline

INF = 10**7

def test_case(n, paths):
    min_cost = [INF] * (n + 1)
    min_cost[1] = 0
    
    for _ in range(n):
        for path in paths:
            st, end, cost = path[0], path[1], path[2]
            
            if min_cost[end] > min_cost[st] + cost:
                min_cost[end] = min_cost[st] + cost
    
    for path in paths:
        st, end, cost = path[0], path[1], path[2]
        
        if min_cost[end] > min_cost[st] + cost:
            return "YES"
            
    return "NO"
    

def solve():
    tc = int(input())
    
    while tc > 0:
        n, m, w = map(int, input().split())
        edges = []
        for _ in range(m):
            a, b, c = map(int, input().split())
            edges.append([a, b, c])
            edges.append([b, a, c])
        for _ in range(w):
            edges.append(list(map(lambda x: -x[1] if x[0] == 2 else x[1], 
                                  enumerate(map(int, input().split())))))
                                  
        print(test_case(n, edges))
        tc -= 1

if __name__ == '__main__':
    solve()

File: Shortest Path/Bellman-Ford/1865번 - 웜홀/1865.cpp
#include <iostream>
#include <vector>

using namespace std;

#define INF 1e9

int t, n, m, w;
vector<vector<int> > edges;
int dist[501];

void input() {
    cin >> n >> m >> w;
    int st, ed, cost;
    
    edges.clear();
    for(int i = 0; i < m; i++) {
        cin >> st >> ed >> cost;
        edges.push_back({st, ed, cost});
        edges.push_back({ed, st, cost});
    }
    for(int i = 0; i < w; i++) {
        cin >> st >> ed >> cost;
        edges.push_back({st, ed, -cost});
    }
}

string bellmanford(int start) {
    dist[start] = 0;
    
    for(int i = 0; i < n; i++) {
        for(auto &edge : edges) {
            int cur_node = edge[0], next_node = edge[1], edge_cost = edge[2];
            
            if(dist[cur_node] + edge_cost < dist[next_node]) {
                dist[next_node] = dist[cur_node] + edge_cost;
            }
        }
    }
    
    for(auto &edge : edges) {
        int cur_node = edge[0], next_node = edge[1], edge_cost = edge[2];
        
        if(dist[cur_node] + edge_cost < dist[next_node]) {
            return "YES";
        }
    }
    
    return "NO";
}

void solve() {
    fill_n(&dist[0], 501, INF);
    cout << bellmanford(1) << '\n';
}

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    
    cin >> t;
    while(t--) {
        input();
        solve();
    }
    
    return 0;
}